### PR TITLE
ENH: stats: another batch of `marray` support

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
gh-22827, gh-22490

#### What does this implement/fix?
Tests `marray` support for `stats.iqr`, `stats.median_abs_deviation`, `stats.expectile` (no modifications required) 
Adds support to `stats.linregress`, `stats.alexandergovern`, `stats.levene`

#### Additional information
The only modifications required were the usual ones:
- The length of an array along the working axis is no longer `.shape[-1]`; use `_length_nonmasked`
- For paired-sample statistics, the masks need to be shared (`_share_masks`)

@steppi With this PR, the coverage for MArray is similar to that for CuPy (because `rankdata` is the main thing holding up both). How can we add it to the Array API support table?

### CODE REVIEW

File: scipy/stats/_morestats.py [hunk 1]
```diff
@@ -3380,12 +3382,15 @@ def func(x):
...
+            message = "`center='trimmed'` is incompatible with MArray."
+            raise ValueError(message)
 
         def func(x):
-            # keepdims=True doesn't currently work for lazy arrays
+            # keepdims=True doesn't currently work for Dask
             return _stats_py.trim_mean(x, proportiontocut, axis=-1)[..., xp.newaxis]
 
-    Nis = [sample.shape[-1] for sample in samples]
+    Nis = [_length_nonmasked(sample, axis=-1, keepdims=True, xp=xp) for sample in samples]
```
Comment (mdhaber) [inline scipy/stats/_morestats.py]:
```suggestion
    Nis = [_length_nonmasked(sample, axis=-1, keepdims=True, xp=xp)
           for sample in samples]
```

File: scipy/stats/tests/test_marray.py [hunk 2]
```diff
@@ -52,7 +52,9 @@ def get_arrays(n_arrays, *, dtype='float64', xp=np, shape=(7, 8), seed=849121654
 @skip_backend('torch', reason="marray#99")
 @pytest.mark.parametrize('fun, kwargs', [make_xp_pytest_param(stats.gmean, {}),
                                          make_xp_pytest_param(stats.hmean, {}),
```
Comment (mdhaber) [inline scipy/stats/tests/test_marray.py]:
```suggestion
@pytest.mark.parametrize('fun, kwargs', [
    make_xp_pytest_param(stats.gmean, {}),
    make_xp_pytest_param(stats.hmean, {}),
```

File: scipy/stats/tests/test_marray.py [hunk 3]
```diff
@@ -52,7 +52,9 @@ def get_arrays(n_arrays, *, dtype='float64', xp=np, shape=(7, 8), seed=849121654
...
-                                         make_xp_pytest_param(stats.pmean, {'p': 2})])
+                                         make_xp_pytest_param(stats.pmean, {'p': 2}),
+                                         make_xp_pytest_param(stats.expectile, {'alpha': 0.4})
+                                         ])
```
Comment (mdhaber) [inline scipy/stats/tests/test_marray.py]:
```suggestion
    make_xp_pytest_param(stats.pmean, {'p': 2}),
    make_xp_pytest_param(stats.expectile, {'alpha': 0.4})
])
```

File: scipy/stats/_morestats.py [hunk 4]
```diff
@@ -3269,7 +3269,9 @@ def bartlett(*samples, axis=0):
 
 
 @xp_capabilities(cpu_only=True, exceptions=['cupy'],
-                 jax_jit=False)  # needs fdtrc
+                 jax_jit=False,  # needs fdtrc
+                 extra_note="Option `center='trimmed' is incompatible with MArray."
```
Comment (j-bowhay) [inline scipy/stats/_morestats.py]:
```suggestion
                 extra_note="Option ``center='trimmed'`` is incompatible with MArray."
```

Comment (mdhaber) [reply scipy/stats/_morestats.py]: Oops, forgot *this* appearance was going on the docs. Will commit when there's something else.

**Review by rkern — APPROVED** [review]
I've given this only a light review and not a deep one, but LGTM! As you said on the Discourse, the risks are small, and I think you took good care here. The CI failure is unrelated.

File: scipy/stats/tests/test_marray.py [hunk 5]
```diff
@@ -375,6 +383,29 @@ def test_pearsonr(f, xp):
         xp_assert_close(res_ci_low.data, xp.asarray(ref_ci_low))
         xp_assert_close(res_ci_high.data, xp.asarray(ref_ci_high))
 
+
+@skip_backend('dask.array', reason='Arrays need `device` attribute: dask/dask#11711')
+@skip_backend('jax.numpy', reason="JAX doesn't allow item assignment.")
+@skip_backend('torch', reason="array-api-compat#242")
+@skip_backend('cupy', reason="special functions won't work")
+@pytest.mark.parametrize('f', [make_xp_pytest_param(stats.linregress),
+                               ])
```
Comment (lucascolley) [inline scipy/stats/tests/test_marray.py]:
```suggestion
@pytest.mark.parametrize('f', [make_xp_pytest_param(stats.linregress)])
```

File: scipy/stats/_stats_py.py
Comment (lucascolley) [inline scipy/stats/_stats_py.py:10648]: could you briefly explain why the special case is needed/acceptable here?

**Review by lucascolley — APPROVED** [review]

File: scipy/stats/_stats_py.py
Comment (mdhaber) [reply scipy/stats/_stats_py.py:10648]: Because `n` is an array for MArrays, so we can't use the code as written.

... And the special treatment here is just wrong. This code is basically saying that a sample with a single observation should have a t-test p-value and SEM of 0. The typical interpretation of such a result is basically the opposite of what it should be.

After this PR, only MArrays will produce the appropriate result in these cases, which is NaN[^1]. I don't think I should fix it in this PR, but I should probably fix it for other backends in a follow-up.

[^1]: There is a permutation argument that `prob` should be defined and `1.0` when the xs and ys are distinct, but that's essentially the opposite of what is done here.

**Comment by lucascolley** [comment]
thanks all
